xds/src/main/java/io/grpc/xds/XdsPicker.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Objects;
+import io.grpc.ConnectivityState;
+import io.grpc.LoadBalancer.PickResult;
+import io.grpc.LoadBalancer.PickSubchannelArgs;
+import io.grpc.LoadBalancer.SubchannelPicker;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+class XdsPicker extends SubchannelPicker {
+
+  private static final WrrAlgorithm wrrAlgo = new WrrAlgorithmImpl();
+
+  private final List<LocalityState> wrrList;
+  private final WrrAlgorithm wrrAlgorithm;
+  private final Map<Locality, SubchannelPicker> childPickers;
+
+  XdsPicker(List<LocalityState> wrrList, Map<Locality, SubchannelPicker> childPickers) {
+    this(wrrList, childPickers, wrrAlgo);
+  }
+
+  @VisibleForTesting
+  XdsPicker(
+      List<LocalityState> wrrList,
+      Map<Locality, SubchannelPicker> childPickers,
+      WrrAlgorithm wrrAlgorithm) {
+    checkArgument(!checkNotNull(wrrList, "wrrList in null").isEmpty(), "wrrList is empty");
+    for (LocalityState localityState : wrrList) {
+      checkArgument(
+          childPickers.containsKey(localityState.locality),
+          "childPickers does not contain picker for locality %s",
+          localityState.locality);
+    }
+    this.wrrList = wrrList;
+    this.childPickers = childPickers;
+    this.wrrAlgorithm = wrrAlgorithm;
+  }
+
+  @Override
+  public PickResult pickSubchannel(PickSubchannelArgs args) {
+    Locality locality = wrrAlgorithm.pickLocality(wrrList);
+    return childPickers.get(locality).pickSubchannel(args);
+  }
+
+  @VisibleForTesting // Interface of weighted Round-Robin algorithm that is convenient for test.
+  interface WrrAlgorithm {
+    Locality pickLocality(List<LocalityState> wrrList);
```

Comment:
This interface doesn't make sense. The wrrList should be passed once (for example at construction time).

Reply:
Done.